src/formatters/junit.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright JS Foundation and other contributors, https://js.foundation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @fileoverview jUnit Reporter
+ * @author Jamund Ferguson
+ */
+
+import { extname } from '@stoplight/path';
+import { DiagnosticSeverity, Dictionary } from '@stoplight/types';
+import { escapeRegExp } from 'tslint/lib/utils';
+import { IRuleResult, SpectralDiagnosticSeverity } from '../types';
+import { Formatter } from './types';
+import { groupBySource } from './utils/groupBySource';
+import { xmlEscape } from './utils/xmlEscape';
+
+const SEVERITY_MAP: Dictionary<string, SpectralDiagnosticSeverity> = {
+  [DiagnosticSeverity.Error]: 'Error',
+  [DiagnosticSeverity.Warning]: 'Warning',
+  [DiagnosticSeverity.Hint]: 'Hint',
+  [DiagnosticSeverity.Information]: 'Information',
+  [-1]: 'off',
+};
+
+function getMessageType(result: IRuleResult) {
+  return SEVERITY_MAP[result.severity];
+}
+
+export const junit: Formatter = results => {
+  let output = '';
+
+  output += '<?xml version="1.0" encoding="utf-8"?>\n';
+  output += '<testsuites>\n';
+
+  const groupedResults = groupBySource(results);
+
+  for (const [source, validationResults] of Object.entries(groupedResults)) {
+    const classname = source.replace(new RegExp(`${escapeRegExp(extname(source))}$`), '');
+
+    if (validationResults.length > 0) {
+      output += `<testsuite package="org.spectral" time="0" tests="${validationResults.length}" errors="${
+        validationResults.length
+      }" name="${source}">\n`;
+
+      for (const result of validationResults) {
+        output += `<testcase time="0" name="org.spectral.${result.code || 'unknown'}" classname="${classname}">`;
+        output += `<error message="${xmlEscape(result.message)}">`;
```

Comment:
One question, is it right that we are calling everything an error? Are there ways to highlight warning or info too? 

I did a very quick google and found this page:

```
<?xml version="1.0" encoding="UTF-8" ?> 
   <testsuites id="20140612_170519" name="New_configuration (14/06/12 17:05:19)" tests="225" failures="1262" time="0.001">
      <testsuite id="codereview.cobol.analysisProvider" name="COBOL Code Review" tests="45" failures="17" time="0.001">
         <testcase id="codereview.cobol.rules.ProgramIdRule" name="Use a program name that matches the source file name" time="0.001">
            <failure message="PROGRAM.cbl:2 Use a program name that matches the source file name" type="WARNING">
WARNING: Use a program name that matches the source file name
Category: COBOL Code Review – Naming Conventions
File: /project/PROGRAM.cbl
Line: 2
      </failure>
    </testcase>
  </testsuite>
</testsuites>
```

They have failures with a type. I don't know if this is the same version of JUnit or anything, just a real quick search, but if we can find a way to show severity somehow that'd be a bonus, because people _will_ ask.

Reply:
You are right. I'll try to include it somehow.